Add notes for #20833 (#300)
bitcoin-core-review-club/website: _posts/2021-01-27-#20833.md

---
layout: pr
date: 2021-01-27
title: "Enable packages through testmempoolaccept"
pr: 20833 
authors: [glozow]
components: ["rpc/rest/zmq", "tests", "validation"]
host: glozow
status: upcoming
commit: 6574758
---

> ## Definitions

Reviewer: nit, maybe add an empty line before this one

- A **package** is a group of transactions that have some dependency relationship, i.e., some
transaction spends the outputs of another transaction in the package.
We might call a package a "transaction chain" when it has multiple generations of such dependencies.

- A **test accept** of validation refers to an invocation of `MemPoolAccept` that validates but
does not submit a transaction to the mempool, called from the `testmempoolaccept` RPC.

- **Coins** ([`CCoins` class](https://doxygen.bitcoincore.org/struct_c_coin.html)) are UTXOs.

> Note that, in our context, coins don't necessarily need to exist on-chain; they can be outputs from

Reviewer: ```suggestion
In our context, coins don't necessarily need to exist on-chain; they can be outputs from
```

transactions in the mempool or some other valid but unconfirmed transaction.
A **CoinsView** is a data structure that contains information about the available coins.

## Background

- The `testmempoolaccept` RPC was introduced in
[PR#11742](https://github.com/bitcoin/bitcoin/pull/11742) and allows users to test raw transactions
for validity.

> On master, `testmempoolaccept` is limited to one transaction at a time, which means users cannot

Reviewer: maybe s/On master/Currently/

test transactions that depend on outputs that the mempool has not seen yet.

> The obstacle is that our validation code uses a `CoinsViewMemPool`, which tracks the coins

Reviewer: The obstacle -> the reason for the limitation (of one transaction at a time)?

available in the current chain state and mempool (and not anything else), to validate transaction
inputs.

- Another reason package mempool accept may be of interest is improving fee bumping.
Currently, Child Pays for Parent does not work for transactions with a feerate below the mempool
minimum fee, which can be dangerous for recipients of time-sensitive transactions.
[PR#20833](https://github.com/bitcoin/bitcoin/pull/20833) is a first step towards package

> validation and hopefully, eventually, Package Relay.

Reviewer: Suggestion: link to https://bitcoincore.reviews/16401, https://github.com/bitcoin/bitcoin/issues/14895 and https://gist.github.com/sdaftuar/8756699bfcad4d3806ba9f3396d4e66a

Author: I'll link to the #16401 review club and optech topic package relay?

To do so, it applies consensus and
[policy](https://github.com/bitcoin/bitcoin/tree/c831e105/src/policy) rules.
Importantly, if a transaction is invalid, `MemPoolAccept` is designed such that it fails as quickly
as possible to mitigate DoS risks.

- Each `MemPoolAccept` instance creates a `CoinsView` from the current chain tip and mempool.
`CCoinsView` is a class with a hierarchical structure (see diagram below).

```
             +--------------------+
             |                    |
             |     CCoinsView     |
             |                    |
             +----------+---------+
                        ^
                        |
             +----------+---------+
             |                    |
             |  CCoinsViewBacked  |
             |                    |
             +------+------+------+
                    ^      ^
           +--------+      +--------+
           |                        |
+----------+---------+  +-----------+---------+
|                    |  |                     |
|  CCoinsViewCache   |  |  CCoinsViewMemPool  |
|                    |  |                     |
+--------------------+  +---------------------+
```

> - `CoinsViewBacked`s can be "layered" by setting one's `Backend` to another. For example, when a

Reviewer: maybe
- replace the "s" after `CoinsViewBacked` with a noun for the type of object or data structure
- explain what `Backend` is before using it (and how it can be set to another one, or what that does or what it means)

`CoinsViewBacked.GetCoin()` cannot find a coin in its own data structures, it may call its
`Backend.GetCoin()`.

- Since the UTXO set is quite large and stored on disk, Bitcoin Core uses a cache to improve
validation speed.
To prevent cache thrashing caused by an attacker broadcasting many invalid transactions spending a
wide variety of UTXOs (a cheap DoS attack), we keep track of which coins are brought in from disk
and uncache them if the transaction turns out to be invalid.

- Validation roughly consists of `PreChecks`, `PolicyScriptChecks`, `ConsensusScriptChecks`, and
`Finalize`.
We've touched on these functions in previous review clubs, e.g.
[#19339](https://bitcoincore.reviews/19339).

> [This Stack Exchange

Reviewer: ```suggestion
This [Stack Exchange
```

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?

2. What are some use cases for enabling packages through `testmempoolaccept`?

3. From a code niceness standpoint, what is the benefit of moving the validation state, replaced
transactions, and fee to `Workspace`?
Should we add `coins_to_uncache` to `Workspace` and make `ATMPArgs` a const? Why or why not?

4. Why do we run all `PreChecks` before script checks? Would it still be correct if we didn't do
this?

5. In
[AcceptSingleTransaction](https://github.com/bitcoin/bitcoin/blob/574758e605e83d0219fbdb80a1bed4bf4ec71e2a/src/validation.cpp#L1037) (with `test_accept=true`), both `PolicyScriptChecks` and
`ConsensusScriptChecks` are run.
However, in
[AcceptMultipleTransactions](https://github.com/bitcoin/bitcoin/blob/574758e605e83d0219fbdb80a1bed4bf4ec71e2a/src/validation.cpp#L1068), we don't run `ConsensusScriptChecks`. Why is that okay?
(Followup question: why do we run `ConsensusScriptChecks` at all in ATMP?)

> 6. When we're fetching a coin from `m_viewmempool`, what 4 places to we look and in what order? Why

Reviewer: s/to/do/

Author: Oops, good catch

> do we check `cache_package_remove` before `cache_package_add` in `CCoinsViewMemPool::GetCoin`?.

Reviewer: ```suggestion
do we check `cache_package_remove` before `cache_package_add` in `CCoinsViewMemPool::GetCoin`?
```

7. If a block arrives while we're running `PreChecks`, do we have to worry about our `CoinsView`
becoming inconsistent?

8. Why don't we need to clear the `cache_package_add` and `cache_package_remove` in
`CCoinsViewMemPool` between multiple calls to `ProcessNewPackage`?

9. What do you think of commit
[`e0a8e1f`](https://github.com/bitcoin/bitcoin/pull/20833/commits/e0a8e1fa9015eea685cb63e68ad1ec3da893fcd9), which removes the line switching the `m_view` backend to `m_dummy`?
What kind of bug is it trying to prevent and how else are we trying to prevent such a bug?

> 10. What are your thoughts on the results API (array of results for each tx, first one if failed)?

Reviewer: ```suggestion
10. What are your thoughts on the proposed `results` object in the RPC output (an array of results for each transaction, or the first one if failed)?
```

Author: Yeah it's weird because I'm proposing something to not change the current API 😅

---
layout: pr
date: 2021-01-27
title: "Enable packages through testmempoolaccept"
pr: 20833
authors: [glozow]
components: ["rpc/rest/zmq", "tests", "validation"]
host: glozow
status: upcoming
commit: ebca80d
---

## Definitions

- A **package** is a group of transactions that have some dependency
  relationship, i.e., some transaction spends the outputs of another
  transaction in the package.  We might call a package a "transaction chain" when

>   it has multiple generations of such dependencies.

Reviewer 2: I'd omit this second sentence, since I think it probably makes things more confusing rather than clarifying:

- a package might not just be a single chain (i.e. each parent only has one child/each child only has one parent)
- we already have chains in Bitcoin :)

Author: Oopsie. Yeah, I guess it's more like "I sometimes say transaction chain"

- A **test accept** of validation refers to an invocation of mempool validation
  that validates but does not submit a transaction to the mempool, called from
  the `testmempoolaccept` RPC.

> - **Coins** ([`CCoins`](https://doxygen.bitcoincore.org/struct_c_coin.html))

Reviewer 2: Maybe s/**Coins**/A **Coin**/ and s/CCoins/CCoin

  are UTXOs.  In our context, coins don't necessarily need to exist on-chain;
  they can be outputs from transactions in the mempool or some other valid but
  unconfirmed transaction.  A **Coins View** is a data structure that contains
  the available coins. The typical usage is to query a Coins View with a
  `COutPoint`.

## Background

- The `testmempoolaccept` RPC was introduced in
  [#11742](https://github.com/bitcoin/bitcoin/pull/11742) and allows users to
  test raw transactions for validity. While the return object is an array (with
  the intention to allow multiple transactions), `testmempoolaccept` is currently
  limited to one transaction at a time, which means users cannot test
  transactions that depend on outputs that the mempool has not seen yet. The main
  reason for this limitation is that our validation code only uses the coins
  available in the current chain state and mempool (and not anything else), to
  validate transaction inputs.

> - Another reason package mempool accept may be of interest is to improve fee

Reviewer 2: 'Another' from what? I think this is the only reason you give.

  recipients of time-sensitive transactions.
  [#20833](https://github.com/bitcoin/bitcoin/pull/20833) is a first step towards
  package validation and hopefully, eventually, [Package
  Relay](https://bitcoinops.org/en/topics/package-relay/).  We have discussed
  Package Relay in a previous review club,
  [#16401](https://bitcoincore.reviews/16401). 

- You can read more about Package Mempool Acceptance
  [here](https://brink.dev/blog/2021/01/21/fellowship-project-package-accept/).

## Notes

- [#20833](https://github.com/bitcoin/bitcoin/pull/20833) enables dry runs of
  package validation.  It does this by having `CCoinsViewMemPool` keep track of
  the Coins created and spent by transactions in the package.

- The
  [`MemPoolAccept`](https://github.com/bitcoin/bitcoin/blob/7f653c3b/src/validation.cpp#L451)
  class represents a "validation session" with the current mempool and chain
  state; it holds `cs_main` and `pool.cs` mutexes throughout validation.  Its job
  is to figure out whether a transaction (received from peers or clients like
  RPC) is a valid candidate for submission to the mempool.  To do so, it applies
  consensus and

>   [policy](https://github.com/bitcoin/bitcoin/tree/c831e105/src/policy) rules.

Reviewer 2: I'm not sure if I'd link to this for policy. There are more policy rules than exist in the /src/policy directory.

  Importantly, if a transaction is invalid, `MemPoolAccept` is designed such that
  it fails as quickly as possible to mitigate DoS risks.

- Each `MemPoolAccept` instance creates a Coins View from the current chain tip
  and mempool.  `CCoinsView` is a class with a hierarchical structure (see
  diagram below).

```
             +--------------------+
             |                    |
             |     CCoinsView     |
             |                    |
             +----------+---------+
                        ^
                        |
             +----------+---------+
             |                    |
             |  CCoinsViewBacked  |
             |                    |
             +------+------+------+
                    ^      ^
           +--------+      +--------+
           |                        |
+----------+---------+  +-----------+---------+
|                    |  |                     |
|  CCoinsViewCache   |  |  CCoinsViewMemPool  |
|                    |  |                     |

> +--------------------+  +---------------------+

Reviewer 2: Nice diagram!

Perhaps it'd be even clearer if we used this:

![image](https://user-images.githubusercontent.com/1063656/105534906-146f3d00-5ce6-11eb-9b05-734636d4a969.png)

(With @jamesob's permission)

Also maybe link to http://localhost:4000/17487, which was a good review club meeting on the coins classes.

Author: Are you saying we have jamesob's permission, or I should ask?

Author: oo also, we don't have `AcceptToMemoryPoolWorker` anymore btw

Reviewer 2: Ah ok, maybe just use the existing diagram.

> Are you saying we have jamesob's permission, or I should ask?

That was unclear! I meant we should ask, but that's moot now :)

  by an attacker broadcasting many invalid transactions spending a wide variety
  of UTXOs (a cheap DoS attack), we keep track of which Coins are brought in from
  disk and uncache them if the transaction turns out to be invalid.

- Validation roughly consists of `PreChecks`, `PolicyScriptChecks`,
  `ConsensusScriptChecks`, and `Finalize`.  We've touched on these functions in
  previous review clubs, e.g.  [#19339](https://bitcoincore.reviews/19339).  This
  [Stack Exchange
  post](https://bitcoin.stackexchange.com/questions/100317/what-is-the-difference-between-policy-and-consensus-when-it-comes-to-a-bitcoin-c)
  about consensus vs policy rules may also be helpful.

- In an effort to not make any API-breaking changes, #20833 returns a `results`
  array with one result per transaction in the same order they were passed in
  or the first failure result.

## Questions

1. Did you review the PR? [Concept ACK, approach ACK, tested ACK, or
   NACK](https://github.com/bitcoin/bitcoin/blob/master/CONTRIBUTING.md#peer-review)?

2. What are some use cases for enabling packages through `testmempoolaccept`?

3. What are some limitations of this implementation of package validation?

> 4. From a code niceness standpoint, what is the benefit of moving the

Reviewer 2: 'niceness' :laughing: 

Maybe s/niceness/quality/ ?

Author: Would you believe me if I said I spent like 10 minutes looking for a better word 😅
